File: mythtv/contrib/mythlink.pl
#!/usr/bin/perl -w
#
# $Date$
# $Revision$
# $Author$
#
# license: GPL
# author:  Chris Petersen (based on the ideas in mythlink.sh from Dale Gass)
#
# mythlink.pl
#
#   Creates symlinks to more human-readable versions of mythtv recording
#   filenames.  See --help for instructions.
#
#   Automatically detects database settings from mysql.txt, and loads
#   the mythtv recording directory from the database (code from nuvexport).
#

# Includes
    use DBI;
    use Getopt::Long;
    use File::Path;

# Some variables we'll use here
    our ($dest, $format, $usage, $underscores);
    our ($dformat, $dseparator, $dreplacement, $separator, $replacement);
    our ($db_host, $db_user, $db_name, $db_pass, $video_dir, @files);
    our ($hostname, $dbh, $sh, $sh2, $q, $q2);

# Default filename format
    $dformat = '%T %- %Y-%m-%d, %g-%i %A %- %S';
# Default separator character
    $dseparator = '-';
# Default replacement character
    $dreplacement = '-';

# Provide default values for GetOptions
    $format      = $dformat;
    $separator   = $dseparator;
    $replacement = $dreplacement;

# Load the destination directory, if one was specified
    GetOptions('dest|destination|path=s'    => \$dest,
               'format=s'                   => \$format,
               'separator=s'                => \$separator,
               'replacement=s'              => \$replacement,
               'usage|help|h'               => \$usage,
               'underscores'                => \$underscores
              );

# Print usage
    if ($usage) {
        print <<EOF;
$0 usage:

options:

--dest

    By default, links will be created in the show_names directory inside of the
    current mythtv data directory on this machine.  eg:

    /var/video/show_names/

--format

    default:  $dformat

    \%T = title    (aka show name)
    \%S = subtitle (aka episode name)
    \%R = description
    \%C = category (as reported by grabber)
    \%U = recording group
    \%p = separator character
    \%y = year, 2 digits
    \%Y = year, 4 digits
    \%n = month
    \%m = month, leading zero
    \%j = day of month
    \%d = day of month, leading zero
    \%g = 12-hour hour
    \%G = 24-hour hour
    \%h = 12-hour hour, with leading zero
    \%H = 24-hour hour, with leading zero
    \%i = minutes
    \%s = seconds
    \%a = am/pm
    \%A = AM/PM

    * For end time, prepend an "e" to the appropriate time/date format code
      above; i.e. "\%eG" gives the 24-hour hour for the end time.

    * For original airdate, prepend an "o" to the year, month, or day format
      codes above; i.e. "\%oY" gives the year in which the episode was first
      aired.

    * A suffix of .mpg or .nuv will be added where appropriate.

--separator

    The string used to separate sections of the link name.  Specifying the
    separator allows trailing separators to be removed from the link name and
    multiple separators caused by missing data to be consolidated. Indicate the
    separator character in the format string using either a literal character
    or the \%- specifier.

    default:  '$dseparator'

--replacement

    Characters in the link name which are not legal on some filesystems will
    be replaced with the given character

    illegal characters:  \\ : * ? < > | "

    default:  '$dreplacement'

--underscores

    Replace whitespace in filenames with underscore characters.

    default:  No underscores

--help

    Show this help text.

EOF
        exit;
    }

# Check the separator and replacement characters for illegal characters
    if ($separator =~ /(?:[\/\\:*?<>|"])/) {
        die "The separator cannot contain any of the following characters:  /\\:*?<>|\"\n";
    }
    elsif ($replacement =~ /(?:[\/\\:*?<>|"])/) {
        die "The replacement cannot contain any of the following characters:  /\\:*?<>|\"\n";
    }

# Escape where necessary
    our $safe_sep = $separator;
        $safe_sep =~ s/([^\w\s])/\\$1/sg;
    our $safe_rep = $replacement;
        $safe_rep =~ s/([^\w\s])/\\$1/sg;

# Get the hostname of this machine
    $hostname = `hostname`;
    chomp($hostname);

# Read the mysql.txt file in use by MythTV.
# could be in a couple places, so try the usual suspects
    my $found = 0;
    my @mysql = ('/usr/local/share/mythtv/mysql.txt',
                 '/usr/share/mythtv/mysql.txt',
                 '/etc/mythtv/mysql.txt',
                 '/usr/local/etc/mythtv/mysql.txt',
                 "$ENV{HOME}/.mythtv/mysql.txt",
                 'mysql.txt'
                );
    foreach my $file (@mysql) {
        next unless (-e $file);
        $found = 1;
        open(CONF, $file) or die "Unable to open $file:  $!\n\n";
        while (my $line = <CONF>) {
        # Cleanup
            next if ($line =~ /^\s*#/);
            $line =~ s/^str //;
            chomp($line);
        # Split off the var=val pairs
            my ($var, $val) = split(/\=/, $line, 2);
            next unless ($var && $var =~ /\w/);
            if ($var eq 'DBHostName') {
                $db_host = $val;
            }
            elsif ($var eq 'DBUserName') {
                $db_user = $val;
            }
            elsif ($var eq 'DBName') {
                $db_name = $val;
            }
            elsif ($var eq 'DBPassword') {
                $db_pass = $val;
            }
        # Hostname override
            elsif ($var eq 'LocalHostName') {
                $hostname = $val;
            }
        }
        close CONF;
    }
    die "Unable to locate mysql.txt:  $!\n\n" unless ($found && $db_host);

# Connect to the database
    $dbh = DBI->connect("dbi:mysql:database=$db_name:host=$db_host", $db_user, $db_pass)
        or die "Cannot connect to database: $!\n\n";
    END {
        $dbh->disconnect if ($dbh);
    }

# Find the directory where the recordings are located
    $q = 'SELECT data FROM settings WHERE value="RecordFilePrefix" AND hostname=?';
    $sh = $dbh->prepare($q);
        $sh->execute($hostname) or die "Could not execute ($q):  $!\n\n";
    ($video_dir) = $sh->fetchrow_array;
    $sh->finish;
    die "This host not configured for myth.\n(No RecordFilePrefix defined for $hostname in the settings table.)\n\n" unless ($video_dir);
    die "Recordings directory $video_dir doesn't exist!\n\n" unless (-d $video_dir);
    $video_dir =~ s/\/+$//;

# Double-check the destination
    $dest ||= "$video_dir/show_names";
    print "Destination directory:  $dest\n";

# Create nonexistent paths
    unless (-e $dest) {
        mkpath($dest, 0, 0755) or die "Failed to create $dest:  $!\n";
    }

# Bad paths
    die "$dest is not a directory.\n" unless (-d $dest);

# Delete any old links
    foreach my $file (<$dest/*>) {
        next unless (-l $file);
        unlink $file or die "Couldn't remove old symlink $file:  $!\n";
    }

# Load the files
    opendir(DIR, $video_dir) or die "Can't open $video_dir:  $!\n\n";
    @files = grep /\.(?:mpg|nuv)$/, readdir(DIR);
    closedir DIR;

# Prepare a database queries
    $q  = 'SELECT * FROM recorded WHERE basename=?';
    $q2 = 'SELECT * FROM recorded WHERE chanid=? AND starttime=?';

    $sh  = $dbh->prepare($q);
    $sh2 = $dbh->prepare($q2);

# Create symlinks for the files on this machine
    foreach my $file (@files) {
        next if ($file =~ /^ringbuf/);
    # Info hash
        my %info;
    # New db format? (no errors, since this query will break with mythtv < .19)
        $sh->execute($file);
        %info = %{$sh->fetchrow_hashref};
    # Nope
        unless ($info{'chanid'}) {
        # Pull out the various parts that make up the filename
            my ($channel, $starttime) = $file =~/^(\d+)_(\d{14})[_\.]/i;
        # Found a bad filename?
            unless ($channel) {
                print "Unknown filename format:  $file\n";
                next;
            }
        # Execute the query
            $sh2->execute($channel, $starttime)
                or die "Could not execute ($q2):  $!\n\n";
            %info = %{$sh2->fetchrow_hashref};
        }
    # Unknown file - someday we should report this
        unless ($info{'chanid'}) {
            print STDERR "Unknown file:  $file\n";
            next;
        }
        my ($title, $subtitle, $description, $recgroup, $category, $oad) = $sh->fetchrow_array;
    # Default times
        my ($syear, $smonth, $sday, $shour, $sminute, $ssecond) = $info{'starttime'} =~ /(\d+)-(\d+)-(\d+)\s+(\d+):(\d+):(\d+)/;
        my ($eyear, $emonth, $eday, $ehour, $eminute, $esecond) = $info{'endtime'}   =~ /(\d+)-(\d+)-(\d+)\s+(\d+):(\d+):(\d+)/;
    # Format some fields we may be parsing below
        # Start time
        my $meridian = ($shour > 12) ? 'AM' : 'PM';
        my $hour = ($shour > 12) ? $shour - 12 : $shour;
        if ($hour < 10) {
            $hour = "0$hour";
        }
        elsif ($hour < 1) {
            $hour = 12;
        }
        # End time
        my $emeridian = ($ehour > 12) ? 'AM' : 'PM';
        my $ethour = ($ehour > 12) ? $ehour - 12 : $ehour;
        if ($ethour < 10) {
            $ethour = "0$ethour";
        }
        elsif ($ethour < 1) {
            $ethour = 12;
        }
    # Original airdate
        $info{'originalairdate'} ||= '0000-00-00';
        my ($oyear, $omonth, $oday) = split(/\-/, $info{'originalairdate'}, 3);
    # Build a list of name format options
        my %fields;
        ($fields{'T'} = ($info{'title'}       or '')) =~ s/%/%%/g;
        ($fields{'S'} = ($info{'subtitle'}    or '')) =~ s/%/%%/g;
        ($fields{'R'} = ($info{'description'} or '')) =~ s/%/%%/g;
        ($fields{'C'} = ($info{'category'}    or '')) =~ s/%/%%/g;
        ($fields{'U'} = ($info{'recgroup'}    or '')) =~ s/%/%%/g;
    # Start time
        $fields{'y'} = substr($syear, 2);   # year, 2 digits
        $fields{'Y'} = $syear;              # year, 4 digits
        $fields{'n'} = int($smonth);        # month
        $fields{'m'} = $smonth;             # month, leading zero
        $fields{'j'} = int($sday);          # day of month
        $fields{'d'} = $sday;               # day of month, leading zero
        $fields{'g'} = int($hour);          # 12-hour hour
        $fields{'G'} = int($shour);         # 24-hour hour
        $fields{'h'} = $hour;               # 12-hour hour, with leading zero
        $fields{'H'} = $shour;              # 24-hour hour, with leading zero
        $fields{'i'} = $sminute;            # minutes
        $fields{'s'} = $ssecond;            # seconds
        $fields{'a'} = lc($meridian);       # am/pm
        $fields{'A'} = $meridian;           # AM/PM
    # End time
        $fields{'ey'} = substr($eyear, 2);  # year, 2 digits
        $fields{'eY'} = $eyear;             # year, 4 digits
        $fields{'en'} = int($emonth);       # month
        $fields{'em'} = $emonth;            # month, leading zero
        $fields{'ej'} = int($eday);         # day of month
        $fields{'ed'} = $eday;              # day of month, leading zero
        $fields{'eg'} = int($ethour);       # 12-hour hour
        $fields{'eG'} = int($ehour);        # 24-hour hour
        $fields{'eh'} = $ethour;            # 12-hour hour, with leading zero
        $fields{'eH'} = $ehour;             # 24-hour hour, with leading zero
        $fields{'ei'} = $eminute;           # minutes
        $fields{'es'} = $esecond;           # seconds
        $fields{'ea'} = lc($emeridian);     # am/pm
        $fields{'eA'} = $emeridian;         # AM/PM
    # Original Airdate
        $fields{'oy'} = substr($oyear, 2);  # year, 2 digits
        $fields{'oY'} = $oyear;             # year, 4 digits
        $fields{'on'} = int($omonth);       # month
        $fields{'om'} = $omonth;            # month, leading zero
        $fields{'oj'} = int($oday);         # day of month
        $fields{'od'} = $oday;              # day of month, leading zero
    # Literals
        $fields{'%'}  = '%';
        ($fields{'-'}  = $separator) =~ s/%/%%/g;
    # Make the substitution
        my $keys = join('|', sort keys %fields);
        my $name = $format;
        $name =~ s/(?<!%)(?:%($keys))/$fields{$1}/g;
        $name =~ s/%%/%/g;
    # Some basic cleanup for illegal (windows) filename characters, etc.
        $name =~ tr/\ \t\r\n/ /s;
        $name =~ tr/"/'/s;
        $name =~ s/(?:[\/\\:*?<>|]+\s*)+(?=[^\d\s])/$replacement /sg;
        $name =~ s/[\/\\:*?<>|]/$replacement/sg;
        $name =~ s/(?:(?:$safe_sep)+\s*)+(?=[^\d\s])/$separator /sg;
        $name =~ s/^($safe_sep|$safe_rep|\ )+//s;
        $name =~ s/($safe_sep|$safe_rep|\ )+$//s;
    # Underscores?
        if ($underscores) {
            $name =~ tr/ /_/s;
        }
    # Get a shell-safe version of the filename (yes, I know it's not needed in this case, but I'm anal about such things)
        my $safe_file = $file;
        $safe_file =~ s/'/'\\''/sg;
        $safe_file = "'$safe_file'";
    # Figure out the suffix
        my $out    = `file $safe_file 2>/dev/null`;
        my $suffix = ($out =~ /mpe?g/i) ? '.mpg' : '.nuv';
    # Check for duplicates
        if (-e "$dest/$name$suffix") {
            my $count = 2;
            while (-e "$dest/$name.$count$suffix") {
                $count++;
            }
            $name .= ".$count";
        }
    # Create the link
        symlink "$video_dir/$file", "$dest/$name$suffix"
            or die "Can't create symlink $dest/$name$suffix:  $!\n";
        print "$dest/$name$suffix\n";
    }

    $sh->finish;
    $sh2->finish;

File: mythtv/contrib/user_jobs/mythlink.py
#!/usr/bin/env python
#
# Ported from mythlink.pl @commit a527db90b27cbc45a1fbec81f6dd2e7511e66bd4
#
# Creates symlinks to mythtv recordings using more-human-readable filenames.
# See --help for instructions.
#
# Automatically detects database settings from mysql.txt, and loads
# the mythtv recording directory from the database (code from nuvexport).
#
# @url       $URL$
# @date      $Date$
# @version   $Revision$
# @author    $Author$
# @license   GPL
#

# Includes
#     use DBI
import argparse, os, re, sys
from functools import partial

import MythTV
from MythTV import FileOps
from MythTV.mythproto import check_ipv6, findfile

# 
# # Some variables we'll use here
#     our ($dest, $format, $usage, $underscores, $live, $rename, args.maxlength)
#     our ($chanid, $starttime, $filename)
#     our ($dformat, $dseparator, $dreplacement, $separator, $replacement)
#     our ($db_host, $db_user, $db_name, $db_pass, $video_dir, $verbose)
#     our ($hostname, $dbh, $sh, $q, $count, $base_dir)
# 
# 
# # Provide default values for GetOptions
#     $format      = $dformat
#     $separator   = $dseparator
#     $replacement = $dreplacement
#     args.maxlength   = -1

class MythBEExt( MythTV.MythBE ):
    @FileOps._ProgramQuery('QUERY_RECORDINGS Descending', sorted=True)
    def getRecordingsDesc(self, pg):
        """
        Returns a list of all Program objects which have already recorded
        """
        return pg
    
    
# Load the cli options
parser = argparse.ArgumentParser()
parser.add_argument('--link', '--destination', '--path', action='store', help='''Specify the directory for the links.  If no pathname is given, links will
    be created in the show_names directory inside of the current mythtv data
    directory on this machine.  eg:

    /var/video/show_names/

    WARNING: ALL symlinks within the destination directory and its
    subdirectories (recursive) will be removed.''', dest='dest')
parser.add_argument('--chanid', action='store', help='''Create a link only for the specified recording file. Use with --starttime
    to specify a recording. This argument may be used with the event-driven
    notification system's "Recording started" event or in a post-recording
    user job.''')
parser.add_argument('--starttime', action='store', help='''Create a link only for the specified recording file. Use with --chanid
    to specify a recording. This argument may be used with the event-driven
    notification system's "Recording started" event or in a post-recording
    user job.''')
parser.add_argument('--filename', action='store', help='''Create a link only for the specified recording file. This argument may be
    used with the event-driven notification system's "Recording started" event
    or in a post-recording user job.''')
parser.add_argument('--format', action='store', default='%T %- %Y-%m-%d, %g-%i %A %- %S', help='''%%T   = Title (show name)
    %%S   = Subtitle (episode name)
    %%R   = Description
    %%eS  = Season (leading zero)
    %%eE  = Episode (leading zero)
    %%eP  = Part Number (Suppressed if not present render as per --renderpart)
    %%in  = Internet reference number
    %%C   = Category
    %%Ct  = Category Type
    %%U   = RecGroup
    %%hn  = Hostname of the machine where the file resides
    %%c   = Channel:  MythTV chanid
    %%cn  = Channel:  channum
    %%cc  = Channel:  callsign
    %%cN  = Channel:  channel name
    %%y   = Recording start time:  year, 2 digits
    %%Y   = Recording start time:  year, 4 digits
    %%n   = Recording start time:  month
    %%m   = Recording start time:  month, leading zero
    %%j   = Recording start time:  day of month
    %%d   = Recording start time:  day of month, leading zero
    %%g   = Recording start time:  12-hour hour
    %%G   = Recording start time:  24-hour hour
    %%h   = Recording start time:  12-hour hour, with leading zero
    %%H   = Recording start time:  24-hour hour, with leading zero
    %%i   = Recording start time:  minutes
    %%s   = Recording start time:  seconds
    %%a   = Recording start time:  am/pm
    %%A   = Recording start time:  AM/PM
    %%ey  = Recording end time:  year, 2 digits
    %%eY  = Recording end time:  year, 4 digits
    %%en  = Recording end time:  month
    %%em  = Recording end time:  month, leading zero
    %%ej  = Recording end time:  day of month
    %%ed  = Recording end time:  day of month, leading zero
    %%eg  = Recording end time:  12-hour hour
    %%eG  = Recording end time:  24-hour hour
    %%eh  = Recording end time:  12-hour hour, with leading zero
    %%eH  = Recording end time:  24-hour hour, with leading zero
    %%ei  = Recording end time:  minutes
    %%es  = Recording end time:  seconds
    %%ea  = Recording end time:  am/pm
    %%eA  = Recording end time:  AM/PM
    %%py  = Program start time:  year, 2 digits
    %%pY  = Program start time:  year, 4 digits
    %%pn  = Program start time:  month
    %%pm  = Program start time:  month, leading zero
    %%pj  = Program start time:  day of month
    %%pd  = Program start time:  day of month, leading zero
    %%pg  = Program start time:  12-hour hour
    %%pG  = Program start time:  24-hour hour
    %%ph  = Program start time:  12-hour hour, with leading zero
    %%pH  = Program start time:  24-hour hour, with leading zero
    %%pi  = Program start time:  minutes
    %%ps  = Program start time:  seconds
    %%pa  = Program start time:  am/pm
    %%pA  = Program start time:  AM/PM
    %%pey = Program end time:  year, 2 digits
    %%peY = Program end time:  year, 4 digits
    %%pen = Program end time:  month
    %%pem = Program end time:  month, leading zero
    %%pej = Program end time:  day of month
    %%ped = Program end time:  day of month, leading zero
    %%peg = Program end time:  12-hour hour
    %%peG = Program end time:  24-hour hour
    %%peh = Program end time:  12-hour hour, with leading zero
    %%peH = Program end time:  24-hour hour, with leading zero
    %%pei = Program end time:  minutes
    %%pes = Program end time:  seconds
    %%pea = Program end time:  am/pm
    %%peA = Program end time:  AM/PM
    %%oy  = Original Airdate:  year, 2 digits
    %%oY  = Original Airdate:  year, 4 digits
    %%on  = Original Airdate:  month
    %%om  = Original Airdate:  month, leading zero
    %%oj  = Original Airdate:  day of month
    %%od  = Original Airdate:  day of month, leading zero
    %%%%  = a literal %% character

    * The program start time is the time from the listings data and is not
      affected by when the recording started.  Therefore, using program start
      (or end) times may result in duplicate names.  In that case, the script
      will append a "counter" value to the name.

    * A suffix of .mpg or .nuv will be added where appropriate.

    * To separate links into subdirectories, include the / format specifier
      between the appropriate fields.  For example, "%%T/%%S" would create
      a directory for each title containing links for each recording named
      by subtitle.  You may use any number of subdirectories in your format
      specifier.''')
parser.add_argument('--live', action='store_true', help='Include live tv recordings.')
parser.add_argument('--separator', action='store', default='-', help='''The string used to separate sections of the link name.  Specifying the
    separator allows trailing separators to be removed from the link name and
    multiple separators caused by missing data to be consolidated. Indicate the
    separator character in the format string using either a literal character
    or the %%- specifier.''')
parser.add_argument('--replacement', action='store', default='-', help='''Characters in the link name which are not legal on some filesystems will
    be replaced with the given character

    illegal characters:  \\ : * ? < > | "''')
parser.add_argument('--rename', action='store_true', help='''Rename the recording files back to their default names.  If you had
    previously used mythrename.py to rename files (rather than creating links
    to the files), use this option to restore the file names to their default
    format.

    Renaming the recording files is no longer supported.  Instead, use
    http://www.mythtv.org/wiki/mythfs.py to create a FUSE file system that
    represents recordings using human-readable file names or use mythlink.pl to
    create links with human-readable names to the recording files.''')
parser.add_argument('--maxlength', action='store', type=int, help='''Ensure the link name is length or fewer characters.  If the link name is
    longer than length, truncate the name. Zero or any negative value for
    length disables length checking.

    Note that this option does not take into account the path length, so on a
    file system used by applications with small path limitations (such as
    Windows Explorer and Windows Command Prompt), you should specify a length
    that takes into account characters used by the path to the dest directory.''', default=0)

parser.add_argument('--usage', action='store_true', help='Show this help text.')
parser.add_argument('--underscores', action='store_true', help='Replace whitespace in filenames with underscore characters.')
parser.add_argument('--verbose', action='store_true', help='Print debug info.')
parser.add_argument('--renderpart', action='store', help='''Use the following as template for rendering part numbers. Use %%x as the place where the part number should go e.g. "(Part %%x)" or "pt%%x".''', default='pt%x')
parser.add_argument('--disablereplacement', action='store_true', help='Stops the disabling of the replacement texts')

args = parser.parse_args()
if args.usage:
    parser.print_usage()
    sys.exit()

# Ensure --chanid and --starttime were specified together, if at all
if (args.chanid or args.starttime) and not (args.chanid and args.starttime):
    raise Exception("The arguments --chanid and --starttime must be used together.")

# Ensure --maxlength specifies a reasonable value (though filenames may
# still be useless at such short lengths)
if args.maxlength > 0 and args.maxlength < 19:
    raise Exception("The --maxlength must be 20 or higher.")

# Check the separator and replacement characters for illegal characters
if args.separator is not None and re.match(r'(?:[\/\\:*?<>|"])', args.separator) is not None:
    raise Exception("The separator cannot contain any of the following characters:  /\\:*?<>|\"")
elif args.replacement is not None and re.match(r'(?:[\/\\:*?<>|"])', args.replacement) is not None:
    raise Exception("The replacement cannot contain any of the following characters:  /\\:*?<>|\"")

safe_sep = None
if args.separator is not None:
    safe_sep = re.sub(r'([^\w\s])', r'\1', args.separator)
safe_rep = None
if args.replacement is not None:
    safe_rep = re.sub(r'([^\w\s])', r'\1', args.replacement)
if args.disablereplacement:
    safe_rep = None

# Connect to mythbackend
#Myth = MythTV.MythTV()

# Connect to the database
dbh = MythTV.MythDB()
mbe = MythBEExt( db = dbh )

# Get the hostname of this machine
hostname = dbh.gethostname()

def GetLocalFilename(show):
    file = show.filename if show.filename.startswith('myth://') else \
                    (show.hostname, show.storagegroup, show.filename)
    mode = 'r'
    db   = MythTV.DBCache(dbh)

    reuri = re.compile(\
        'myth://((?P<group>.*)@)?(?P<host>[\[\]a-zA-Z0-9_\-\.]*)(:[0-9]*)?/(?P<file>.*)')
    reip = re.compile('(?:\d{1,3}\.){3}\d{1,3}')

    if mode not in ('r','w'):
        raise TypeError("File I/O must be of type 'r' or 'w'")

    # process URI (myth://<group>@<host>[:<port>]/<path/to/file>)
    match = None
    try:
        match = reuri.match(file)
    except:
        pass

    if match:
        host = match.group('host')
        filename = match.group('file')
        sgroup = match.group('group')
        if sgroup is None:
            sgroup = 'Default'
    elif len(file) == 3:
        host, sgroup, filename = file
    else:
        raise MythTV.MythError('Invalid FileTransfer input string: '+file)

    # get full system name
    host = host.strip('[]')
    if reip.match(host) or check_ipv6(host):
        host = db._gethostfromaddr(host)

    # search for file in local directories
    sg = findfile(filename, sgroup, db)
    if sg is None:
        return None

    return sg.dirname+filename

def FindRecordingDir(basename):
    for sgroup in dbh.getStorageGroup():
        if os.path.exists(os.path.join(sgroup.dirname, basename)):
            return sgroup.dirname

    return ''

# Check the length of the link name
def cut_down_name(name, suffix):
    if args.maxlength > 0:
        charsavailable = args.maxlength - len(suffix)
        if args.charsavailable > 0:
            name = name[0:charsavailable]
        else:
            name = ''
    return name

# Print the message, but only if verbosity is enabled
def vprint(message):
    if not args.verbose:
        return
    print(message)
    
    
def padNr(padBy, toPad):
    if toPad is None:
        return ''
    try:
        return str(format(int(toPad), padBy))
    except:
        return ''

def partNumber(partNumber):
    if partNumber is None or args.renderpart is None or partNumber == 0:
        return ''
    return str(args.renderpart).replace('%x', str(partNumber))
    
def showFormatPathExt(show, path, replace=None):
    for (tag, data, fn) in (('eS','season',partial(padNr, '02')), ('eE','episode', partial(padNr, '02')),
                            ('eP','part_number', partNumber)):
            path = path.replace('%'+tag, fn(show[data]))
    return show.formatPath(path, replace)

# Rename the file back to default format
def do_rename():
    for show in mbe.getRecordingsDesc():
        foundFile = GetLocalFilename(show)
        # File doesn't exist locally
        if foundFile is None or not os.path.isfile(foundFile):
            continue
        
        # Format the name
        name = showFormatPathExt(show, '%c_%Y%m%d%H%i%s')
        
        # Get a shell-safe version of the filename (yes, I know it's not needed in this case, but I'm anal about such things)
        _safe_file = "'" + re.sub('\'', '\\\'', foundFile) + "'"

        # Figure out the suffix
        _filename, suffix = os.path.splitext(foundFile)
        name = name[0:len(name)-len(suffix)]
        basename = os.path.basename(foundFile)
        
        # Rename the file, but only if it's a real file
        if suffix is None or len(suffix) == 0:
            continue
        
        # Check for duplicates
        video_dir = FindRecordingDir(basename)
        if os.path.isfile(os.path.join(video_dir, name + suffix)):
            count = 2
            while os.path.isfile(os.path.join(video_dir, name + "." + str(count) + suffix)):
                count = count + 1
            name += "." + str(count)
        name += suffix
        
        originalFilename = foundFile 
        show.filename = show.filename.replace(basename, name)
        show.update()
        
        try:
            os.rename(originalFilename, os.path.join(video_dir, name))
        except:
            show.filename = originalFilename
            show.update()
            raise

        vprint(originalFilename + "\t-> " + show.filename)
        
        # Rename previews
        for thumb in os.listdir(video_dir):
            if thumb.endswith(".png"):
                match = re.match(r'^' + basename + r'((?:\.\d+)?(?:\.\d+x\d+(?:x\d+)?)?)\.png$', thumb)
                if not match:
                    continue
                dim = match.group(1)
                try:
                    os.rename(os.path.join(video_dir, thumb), os.path.join(video_dir, name + dim + ".png"))
                except:
                    # If the rename fails, try to delete the preview from the
                    # cache (it will automatically be re-created with the
                    # proper name, when necessary)
                    try:
                        os.remove(os.path.join(video_dir, thumb))
                    except:
                        vprint("Unable to rename preview image: '" + os.path.join(video_dir, thumb) + "'.")
                     
# Only if we're renaming files back to "default" names
if args.rename:
    do_rename()
    sys.exit()

# Get our base location
base_dir = FindRecordingDir('show_names')

if base_dir == '':
    base_dir = dbh.getStorageGroup().next().dirname

# Link destination
# Double-check the destination
dest = args.dest

# Alert the user
if dest is None:
    dest = os.path.join(base_dir, 'show_names')

vprint("Link destination directory: " + dest)

# Create nonexistent paths
if dest is not None and not os.path.exists(dest):
    os.makedirs(dest)

# Bad path
if dest is None or not os.path.isdir(dest):
    raise Exception(str(dest) + " is not a directory.")

# Delete old links/directories unless linking only one recording
if args.filename and args.chanid:
    for root, dirnames, filenames in os.walk(dest):
        for filename in filenames:
            if os.path.islink(os.path.join(root, filename)):
                os.remove(os.path.join(root, filename))
            
    # Delete empty directories (should this be an option?)
    # Let this fail silently for non-empty directories
    for root, dirnames, filenames in os.walk(dest, topdown=False):
        for dirname in dirnames:
            os.rmdir(os.path.join(dirname, filename))

# Create symlinks for the files on this machine
rows = None
if args.chanid:
    rows = [ mbe.getRecording(args.chanid, args.starttime) ]
else:
    rows = mbe.getRecordingsDesc()
    
for show in rows:
    foundFile = GetLocalFilename(show)

    # File doesn't exist locally
    if foundFile is None or not os.path.exists(foundFile):
        continue
    
    # Skip LiveTV recordings?
    if not (args.live or show.recgroup != 'LiveTV'):
        continue

    # Check if this is the file to link if only linking one file
    if args.filename and not (os.basename(foundFile) == args.filename or
                     foundFile == args.filename):
        continue
    elif args.chanid:
        if show.chanid != args.chanid:
            continue
        recstartts = show.recstartts
        
        # Check starttime in MythTV time format (yyyy-MM-ddThh:mm:ss)
        if recstartts != args.starttime:
            # Check starttime in ISO time format (yyyy-MM-dd hh:mm:ss)
            recstartts = recstartts.replace('T', ' ')
            if recstartts != args.starttime:
                # Check starttime in job queue time format (yyyyMMddhhmmss)
                recstartts = re.sub(r'[\- :]', '', recstartts)
                if recstartts != args.starttime:
                    continue
                
    # Format the name
    name = showFormatPathExt(show, args.format,safe_rep)
    
    # Get a shell-safe version of the filename (yes, I know it's not needed in this case, but I'm anal about such things)
    _safe_file = "'" + re.sub('\'', '\\\'', foundFile) + "'"
    # Figure out the suffix
    _filename, suffix = os.path.splitext(foundFile)
    name = name[0:len(name)-len(suffix)]
    basename = os.path.basename(foundFile)
    
    # Check the link name's length
    name = cut_down_name(name, suffix)
    # Link destination
    # Check for duplicates
    if name and os.path.isfile(os.path.join(dest, name + suffix)):
        if ((args.filename and not args.chanid) or (not os.path.islink(os.path.join(dest, name + suffix)))):
            count = 2
            name = cut_down_name(name, "." + str(count) + suffix)
            while name and os.path.isfile(os.path.join(dest, name + "." + str(count) + suffix)): 
                count = count + 1
                name = cut_down_name(name, "." + str(count) + suffix)
            if name:
                name += "." + str(count)
        else:
            os.remove(os.path.join(dest, name + suffix))
        
    if not name:
        vprint("Unable to represent recording; maxlength too small.")
        continue
    
    name += suffix

    # Create the link
    directory = os.path.dirname(os.path.join(dest, name))
    if not os.path.isdir(directory):
        os.makedirs(directory)
        
    os.symlink(foundFile, os.path.join(dest, name))
    vprint(os.path.join(dest, name))
    
